statsmodels/tsa/statespace/mlemodel.py

```diff
@@ -0,0 +1,797 @@
+            Called after each iteration, as callback(xk), where xk is the
+            current parameter vector.
+        return_params : boolean, optional
+            Whether or not to return only the array of maximizing parameters.
+            Default is False.
+        bfgs_tune : boolean, optional
+            BFGS methods by default use internal methods for approximating the
+            score and hessian by finite differences. If `bfgs_tune=True` the
+            maximizing parameters from the BFGS method are used as starting
+            parameters for a second round of maximization using complex-step
+            differentiation. Has no effect for other methods. Default is False.
+
+        Returns
+        -------
+        MLEResults
+
+        See also
+        --------
+        statsmodels.base.model.LikelihoodModel.fit : for more information
+            on using the solvers.
+        MLEResults : results class returned by fit
+        """
+
+        if start_params is None:
+            start_params = self.start_params
+            transformed = True
+
+        # Unconstrain the starting parameters
+        if transformed:
+            start_params = self.untransform_params(np.array(start_params))
+
+        if method == 'lbfgs' or method == 'bfgs':
+            kwargs.setdefault('approx_grad', True)
+            kwargs.setdefault('epsilon', 1e-5)
+
+        # Maximum likelihood estimation
+        # Set the optional arguments for the loglikelihood function to
+        # maximize the average loglikelihood, by default.
+        fargs = (kwargs.get('average_loglike', True), False, False)
+        mlefit = super(MLEModel, self).fit(start_params, method=method,
+                                           fargs=fargs,
+                                           maxiter=maxiter,
+                                           full_output=full_output, disp=disp,
+                                           callback=callback, **kwargs)
+
+        # Optionally tune the maximum likelihood estimates using complex step
+        # gradient
+        if bfgs_tune and method == 'lbfgs' or method == 'bfgs':
+            kwargs['approx_grad'] = False
+            del kwargs['epsilon']
+            fargs = (kwargs.get('average_loglike', True), False, False)
+            mlefit = super(MLEModel, self).fit(mlefit.params, method=method,
+                                               fargs=fargs,
+                                               maxiter=maxiter,
+                                               full_output=full_output,
+                                               disp=disp, callback=callback,
+                                               **kwargs)
+
+        # Constrain the final parameters and update the model to be sure we're
+        # using them (in case, for example, the last time update was called
+        # via the optimizer it was a gradient calculation, etc.)
+        self.update(mlefit.params, transformed=False)
```

Comment:
Why are you attaching `params` to the model? Is it used somewhere?
`self.update` also returns the params.

If possible, I'd like to avoid attaching fit results to the model to avoid any chances of stale state.